group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,413 @@
+     * <li> Compute the quotas of partitions for each member based on the total partitions and member count.</li>
+     * <li> Initialize unassigned partitions to all the topic partitions and
+     *      remove partitions from the list as and when they are assigned.</li>
+     * <li> For existing assignments, retain partitions based on the determined quota and member's rack compatibility.</li>
+     * <li> If a partition's rack mismatches with its owner, track it for future use.</li>
+     * <li> Identify members that haven't fulfilled their partition quota or are eligible to receive extra partitions.</li>
+     * <li> Depending on members needing extra partitions, select members from the potentially unfilled list
+     *      and add them to the unfilled list.</li>
+     * <li> Proceed with a round-robin assignment adhering to rack awareness.
+     *      For each unassigned partition, locate the first compatible member from the unfilled list.</li>
+     * <li> If no rack-compatible member is found, revert to the tracked current owner.
+     *      If that member can't accommodate the partition due to quota limits, resort to a generic round-robin assignment.</li>
+     */
+    @Override
+    protected GroupAssignment buildAssignment() throws PartitionAssignorException {
+        int totalPartitionsCount = 0;
+
+        for (Uuid topicId : subscriptionIds) {
+            int partitionCount = subscribedTopicDescriber.numPartitions(topicId);
+            if (partitionCount == -1) {
+                throw new PartitionAssignorException(
+                    "Members are subscribed to topic " + topicId + " which doesn't exist in the topic metadata."
+                );
+            } else {
+                totalPartitionsCount += partitionCount;
+            }
+        }
+
+        if (subscriptionIds.isEmpty()) {
+            LOG.info("The subscription list is empty, returning an empty assignment");
+            return new GroupAssignment(Collections.emptyMap());
+        }
+
+        // The minimum required quota that each member needs to meet for a balanced assignment.
+        // This is the same for all members.
+        final int numberOfMembers = assignmentSpec.members().size();
+        final int minQuota = totalPartitionsCount / numberOfMembers;
+        remainingMembersToGetAnExtraPartition = totalPartitionsCount % numberOfMembers;
+
+        assignmentSpec.members().keySet().forEach(memberId ->
+            targetAssignment.put(memberId, new MemberAssignment(new HashMap<>())
+        ));
+
+        // sorted list of all partitions.
+        unassignedPartitions = allTopicIdPartitions(subscriptionIds, subscribedTopicDescriber);
+
+        assignStickyPartitions(minQuota);
+
+        unfilledMembers = computeUnfilledMembers();
+
+        if (!unassignedPartitionsCountEqualsRemainingAssignmentsCount()) {
+            throw new PartitionAssignorException("Number of available partitions is not equal to the total requirement");
+        }
```

Comment:
Is this really necessary? It is a tad annoying to have to compute the sum of all the unfilled members. If you really want to do this, we could perhaps maintain the total count as we update the unfilled members but I am not sure if it is worth it.

Reply:
I added it as a check to ensure we don't have more partitions to assign and less total required assignments count, i.e. jic our calculations are wrong. Its not completely necessary